plugins/home/src/components/CustomHomepage/CustomHomepageGrid.tsx

```diff
@@ -0,0 +1,345 @@
+        backgroundColor: theme.palette.primary.light,
+      },
+      '&.edit > :active': {
+        cursor: 'move',
+      },
+    },
+  }),
+);
+
+export type Widget = WidgetSettings & {
+  name: string;
+  title?: string;
+  component: ReactNode & {
+    type: any;
+    props: any;
+  };
+};
+
+type WidgetStoreItem = {
+  id: string;
+  layout: Layout;
+  settings: Record<string, any>;
+};
+
+const getAvailableWidgets = (children: ReactNode[]): Widget[] => {
+  const ret = children.map(child => {
+    const componentData = getAllComponentData(child);
+    if (!componentData) {
+      return null;
+    }
+    return {
+      component: child,
+      ...Object.fromEntries(componentData as Iterable<any>),
+    } as Widget;
+  });
+  return compact(ret);
+};
+
+function useHomeStorage(
+  defaultWidgets: WidgetStoreItem[],
+): [WidgetStoreItem[], (value: WidgetStoreItem[]) => void] {
+  const storageApi = useApi(storageApiRef);
+  const homeKey = 'home';
+  // TODO: Support multiple home pages
+  const setWidgets = useCallback(
+    (value: WidgetStoreItem[]) => {
+      storageApi.set(homeKey, JSON.stringify(value));
+    },
+    [homeKey, storageApi],
+  );
+  const homeSnapshot = useObservable(
+    storageApi.observe$<string>(homeKey),
+    storageApi.snapshot(homeKey),
+  );
+  const widgets: WidgetStoreItem[] = useMemo(() => {
+    if (homeSnapshot.presence === 'absent') {
+      return defaultWidgets;
+    }
+    try {
+      return JSON.parse(homeSnapshot.value!);
+    } catch (e) {
+      return defaultWidgets;
+    }
+  }, [homeSnapshot, defaultWidgets]);
+
+  return [widgets, setWidgets];
+}
+
+const convertConfigToDefaultWidgets = (
+  config: LayoutConfiguration[],
+  widgets: any[],
+): WidgetStoreItem[] => {
+  const ret = config.map((c, i) => {
+    const name = getComponentData(c.component, 'name');
+    if (!name) {
+      return null;
+    }
+    const widget = widgets.find(w => w.name === name);
+    if (!widget) {
+      return null;
+    }
+    const widgetId = `${widget.name}__${i}`;
+    return {
+      id: widgetId,
+      layout: {
+        i: widgetId,
+        x: c.x,
+        y: c.y,
+        w: c.width,
+        h: c.height,
+        minW: widget.minWidth,
+        maxW: widget.maxWidth,
+        minH: widget.minHeight,
+        maxH: widget.maxHeight,
+        isDraggable: false,
+        isResizable: false,
+      },
+      settings: {},
+    };
+  });
+  return compact(ret);
+};
+
+/**
+ * Layout configuration that can be passed to the custom home page.
+ *
+ * @public
+ */
+export type LayoutConfiguration = {
+  component: ReactNode;
+  x: number;
+  y: number;
+  width: number;
+  height: number;
+};
+
+/**
+ * A component that allows customizing components in home grid layout.
+ *
+ * @public
+ */
+export const CustomHomepageGrid = (props: {
+  children: ReactNode[];
+  config?: LayoutConfiguration[];
+}) => {
+  const styles = useStyles();
+  const availableWidgets = getAvailableWidgets(props.children);
```

Comment:
Hm this is probably unfortunate. You probably want to use `useElementFilter` which is the canonical way to get provided entities out of your children collection. The `children.map` above looks scary.

See for example [this one](https://github.com/backstage/backstage/blob/5b975591efecc2a91ed5aae4965630aecf8a9634/packages/core-app-api/src/routing/FlatRoutes.tsx#L54) which gets all of them and flatmaps, or more importantly [this one](https://github.com/backstage/backstage/blob/5b975591efecc2a91ed5aae4965630aecf8a9634/plugins/catalog/src/components/EntityLayout/EntityLayout.tsx#L188) which filters by specific component data, which probably is what you want.

If I understood the use case correctly, it may mean that you can get rid of the `getAllComponentData` function which may perhaps be a good thing.

Reply:
Ok, will check this, thanks!

Reply:
Thanks for the pointers, should be much safer now, please check!